lib/src/main/java/graphql/nadel/engine/blueprint/NadelExecutionBlueprintFactory.kt

```diff
@@ -403,46 +403,60 @@ private class Factory(
-
-                val prefixes = paths
-                    .asSequence()
-                    .map {
-                        it.segments.dropLast(1) + "*"
-                    }
-                    .toSet()
-
-                // Say we have paths = [
-                //     [page]
-                //     [page.id]
-                //     [page.status]
-                // ]
-                // (e.g. page was the input and the page.id and page.status are used to match batch objects)
-                // then this maps it to [
-                //     [page.id]
-                //     [page.status]
-                // ]
-                paths
-                    .filter {
-                        !prefixes.contains(it.segments + "*")
-                    }
-            },
+            sourceFields = getBatchHydrationSourceFields(matchStrategy, hydrationArgs, condition),
             condition = condition,
         )
     }
 
-    private fun selectSourceFieldQueryPaths(hydrationValueSource: FieldResultValue): List<NadelQueryPath> {
+    private fun getBatchHydrationSourceFields(
+        matchStrategy: NadelBatchHydrationMatchStrategy,
+        hydrationArgs: List<NadelHydrationActorInputDef>,
+        condition: NadelHydrationCondition?
+    ): List<NadelQueryPath> {
+        val sourceFieldsFromArgs = Unit.let {
+            val paths = (when (matchStrategy) {
+                NadelBatchHydrationMatchStrategy.MatchIndex -> emptyList()
+                is NadelBatchHydrationMatchStrategy.MatchObjectIdentifier -> listOf(matchStrategy.sourceId)
+                is NadelBatchHydrationMatchStrategy.MatchObjectIdentifiers -> matchStrategy.objectIds.map { it.sourceId }
+            } + hydrationArgs.flatMap {
+                when (val hydrationValueSource: NadelHydrationActorInputDef.ValueSource = it.valueSource) {
+                    is NadelHydrationActorInputDef.ValueSource.ArgumentValue -> emptyList()
+                    is FieldResultValue -> selectSourceFieldQueryPaths(hydrationValueSource, condition)
+                    is NadelHydrationActorInputDef.ValueSource.StaticValue -> emptyList()
+                }
+            }).toSet()
+
+            val prefixes = paths
+                .asSequence()
+                .map {
+                    it.segments.dropLast(1) + "*"
+                }
+                .toSet()
+
+            // Say we have paths = [
+            //     [page]
+            //     [page.id]
+            //     [page.status]
+            // ]
+            // (e.g. page was the input and the page.id and page.status are used to match batch objects)
+            // then this maps it to [
+            //     [page.id]
+            //     [page.status]
+            // ]
+            paths
+                .filter {
+                    !prefixes.contains(it.segments + "*")
+                }
+        }
+
+        if (condition != null) {
+            return sourceFieldsFromArgs + condition.fieldPath
+        }
```

Comment:
Ok I think we need to add `condition.fieldPath` to `paths` instead. That way we dedup some paths.

Please get rid of the `Unit.let` too. It was meant to house the code in one block, but it's not needed here.

Reply:
To clarify, we need to do that to dedup the fields in the selection set. I noticed this yesterday when we were pairing.

i.e. it should help with

```
batch_hydration__data__dataId: dataId
batch_hydration__data__dataId: dataId
batch_hydration__data__dataId: dataId
batch_hydration__data__dataId: dataId
```

Reply:
Oh I see, thanks. I've made these changes in the latest commit